File: leetcode/2958. length of longest subarray with at most k frequency.go
func maxSubarrayLength(nums []int, k int) int {
	freq := make(map[int]int)
	maxLen := 0
	left := 0

	for right, value := range nums {
		freq[value] += 1

		for freq[value] > k {
			freq[nums[left]] -= 1
			left += 1
		}

		maxLen = max(maxLen, right+1-left)
	}

	return maxLen
}

func maxSubarrayLength(nums []int, k int) int {
	freq := make(map[int]int)
	left := 0
	right := 0
	value := -1
	countGreaterThanK := 0

	for right, value = range nums {
		freq[value] += 1
		if freq[value] == k+1 {
			countGreaterThanK += 1
		}

		if countGreaterThanK > 0 {
			freq[nums[left]] -= 1
			if freq[nums[left]] == k {
				countGreaterThanK -= 1
			}
			left += 1
		}
	}

	return right + 1 - left
}

File: leetcode/2958. length of longest subarray with at most k frequency.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxSubarrayLength = function (nums, k) {
    const maxFreq = k;
    const len = nums.length;
    const freq = [];
    const hashmap = {};
    let maxLen = 0;

    for (let left = 0, right = 0; right < len; right += 1) {
        freq[nums[right]] = (freq[nums[right]] || 0) + 1;

        while (freq[nums[right]] > maxFreq) {
            freq[nums[left]] -= 1;
            left += 1;
        }
        maxLen = Math.max(maxLen, right + 1 - left);
    }
    return maxLen;
};
